File: legacy/sdk/typescript/src/resources/workflows.ts
/** Workflows resource. */
import type { Axion } from "../client.js";
import { AxionError } from "../errors.js";
import type { WorkflowSpec } from "../workflow.js";
import type { WorkflowRun } from "../types.js";

export class WorkflowsResource {
  constructor(private readonly _c: Axion) {}

  list(opts?: { status?: string }): Promise<unknown[]> {
    return this._c._request("GET", "/workflows", { query: opts });
  }

  get(slug: string, opts?: { version?: number }): Promise<unknown> {
    return this._c._request("GET", `/workflows/${slug}`, { query: opts });
  }

  create(slug: string, spec: WorkflowSpec): Promise<unknown> {
    return this._c._request("POST", "/workflows", { body: { slug, spec } });
  }

  publish(slug: string, opts?: { version?: number }): Promise<unknown> {
    return this._c._request("POST", `/workflows/${slug}/publish`, {
      query: opts,
    });
  }

  run(slug: string, inputs: Record<string, unknown> = {}): Promise<WorkflowRun> {
    return this._c._request<WorkflowRun>("POST", `/workflows/${slug}/run`, {
      body: { inputs },
    });
  }

  getRun(runId: string): Promise<WorkflowRun> {
    return this._c._request<WorkflowRun>("GET", `/workflows/runs/${runId}`);
  }

  getRunSteps(runId: string): Promise<unknown[]> {
    return this._c._request<unknown[]>("GET", `/workflows/runs/${runId}/steps`);
  }

  async cancelRun(runId: string): Promise<void> {
    await this._c._request("POST", `/workflows/runs/${runId}/cancel`);
  }

  replayRun(runId: string): Promise<WorkflowRun> {
    return this._c._request<WorkflowRun>(
      "POST",
      `/workflows/runs/${runId}/replay`
    );
  }

  async waitForRun(
    runId: string,
    opts: { pollIntervalMs?: number; timeoutMs?: number } = {}
  ): Promise<WorkflowRun> {
    const interval = opts.pollIntervalMs ?? 2_000;
    const deadline = Date.now() + (opts.timeoutMs ?? 600_000);
    const terminal = new Set(["succeeded", "failed", "cancelled", "awaiting_approval"]);
    while (Date.now() < deadline) {
      const run = await this.getRun(runId);
      if (terminal.has(run.status)) return run;
      await new Promise((r) => setTimeout(r, interval));
    }
    throw new AxionError(`Run ${runId} did not finish in time`);
  }
}

File: legacy/sdk/python/axion/resources/workflows.py
"""Workflows resource."""
from __future__ import annotations

import time
from typing import TYPE_CHECKING, Any

from axion.errors import AxionError

if TYPE_CHECKING:
    from axion.client import Axion


class WorkflowsResource:
    def __init__(self, client: "Axion") -> None:
        self._c = client

    def list(self, *, status: str | None = None) -> list[dict]:
        params = {"status": status} if status else {}
        return self._c._request("GET", "/workflows", params=params)

    def get(self, slug: str, *, version: int | None = None) -> dict:
        params = {"version": version} if version else {}
        return self._c._request("GET", f"/workflows/{slug}", params=params)

    def create(self, slug: str, spec: dict[str, Any]) -> dict:
        return self._c._request("POST", "/workflows", json={"slug": slug, "spec": spec})

    def publish(self, slug: str, *, version: int | None = None) -> dict:
        params = {"version": version} if version else {}
        return self._c._request("POST", f"/workflows/{slug}/publish", params=params)

    def run(self, slug: str, *, inputs: dict[str, Any] | None = None) -> dict:
        return self._c._request(
            "POST", f"/workflows/{slug}/run", json={"inputs": inputs or {}}
        )

    def get_run(self, run_id: str) -> dict:
        return self._c._request("GET", f"/workflows/runs/{run_id}")

    def get_run_steps(self, run_id: str) -> list[dict]:
        return self._c._request("GET", f"/workflows/runs/{run_id}/steps")

    def cancel_run(self, run_id: str) -> None:
        self._c._request("POST", f"/workflows/runs/{run_id}/cancel")

    def replay_run(self, run_id: str) -> dict:
        return self._c._request("POST", f"/workflows/runs/{run_id}/replay")

    def wait_for_run(
        self,
        run_id: str,
        *,
        poll_interval_s: float = 2.0,
        timeout_s: float = 600.0,
    ) -> dict:
        """Block until the run reaches a terminal state."""
        deadline = time.time() + timeout_s
        terminal = {"succeeded", "failed", "cancelled"}
        while time.time() < deadline:
            run = self.get_run(run_id)
            if run["status"] in terminal:
                return run
            if run["status"] == "awaiting_approval":
                return run  # caller decides
            time.sleep(poll_interval_s)
        raise AxionError(f"Run {run_id} did not finish within {timeout_s}s")
